homeassistant/components/nest/media_source.py

```diff
@@ -0,0 +1,259 @@
+                "Unable to find device with identifier: %s" % item.identifier
+            )
+        events = _get_events(device)
+        if media_id.event_id not in events:
+            raise Unresolvable(
+                "Unable to find event with identifier: %s" % item.identifier
+            )
+        event = events[media_id.event_id]
+        return PlayMedia(
+            EVENT_MEDIA_API_URL_FORMAT.format(
+                device_id=media_id.device_id, event_id=media_id.event_id
+            ),
+            event.event_image_type.content_type,
+        )
+
+    async def async_browse_media(self, item: MediaSourceItem) -> BrowseMediaSource:
+        """Return media for the specified level of the directory tree.
+
+        The top level is the root that contains devices. Inside each device are
+        media for events for that device.
+        """
+        media_id: MediaId | None = parse_media_id(item.identifier)
+        _LOGGER.debug(
+            "Browsing media for identifier=%s, media_id=%s", item.identifier, media_id
+        )
+        devices = await self.devices()
+        if media_id is None:
+            # Browse the root and return child devices
+            browse_root = _browse_root()
+            browse_root.children = []
+            for device_id, child_device in devices.items():
+                browse_root.children.append(
+                    _browse_device(MediaId(device_id), child_device)
+                )
+            return browse_root
+
+        # Browse either a device or events within a device
+        if not (device := devices.get(media_id.device_id)):
+            raise BrowseError(
+                "Unable to find device with identiifer: %s" % item.identifier
+            )
+        if media_id.event_id is None:
+            # Browse a specific device and return child events
+            browse_device = _browse_device(media_id, device)
+            browse_device.children = []
+            events = _get_events(device)
+            for child_event in events.values():
+                event_id = MediaId(media_id.device_id, child_event.event_id)
+                browse_device.children.append(
+                    _browse_event(event_id, device, child_event)
+                )
+            return browse_device
+
+        # Browse a specific event
+        events = _get_events(device)
+        if not (event := events.get(media_id.event_id)):
+            raise BrowseError(
+                "Unable to find event with identiifer: %s" % item.identifier
+            )
+        return _browse_event(media_id, device, event)
+
+    async def devices(self) -> Mapping[str, Device]:
+        """Return all event media related devices."""
+        return await get_media_source_devices(self.hass)
+
+
+def _get_events(device: Device) -> Mapping[str, ImageEventBase]:
+    """Return relevant events for the specified device."""
+    return OrderedDict({e.event_id: e for e in device.event_media_manager.events})
```

Comment:
Is it important to return an OrderedDict and not just a dict, which is ordered by default?

Reply:
I did not know `dict` is ordered by default. I just need the events to be iterable in order, not that an OrderedDict itself is used. Thanks, I'll send a follow up.

Reply:
Ah! This is "new"in python 3.7 (or .6) thanks.

Reply:
👍